cpp/include/cuvs/neighbors/cagra.hpp

```diff
@@ -37,31 +40,45 @@ namespace cuvs::neighbors::cagra {
  */
 
 /**
- * @brief ANN algorithm used by CAGRA to build knn graph
+ * @brief ANN parameters used by CAGRA to build knn graph
  *
  */
-enum class graph_build_algo {
-  /* Use IVF-PQ to build all-neighbors knn graph */
-  IVF_PQ,
-  /* Experimental, use NN-Descent to build all-neighbors knn graph */
-  NN_DESCENT
+struct graph_build_params {
+  bool construct_index_with_dataset = true;
+  virtual ~graph_build_params() {}
 };
 
+namespace ivf_pq {
+/** Specialized parameters utilizing IVF-PQ to build knn graph */
+struct graph_build_params : cuvs::neighbors::cagra::graph_build_params {
+  std::optional<cuvs::neighbors::ivf_pq::index_params> build_params   = std::nullopt;
+  std::optional<cuvs::neighbors::ivf_pq::search_params> search_params = std::nullopt;
+  float refinement_rate                                               = 2.0f;
+};
+}  // namespace ivf_pq
+
+namespace nn_descent {
+/** Specialized parameters utilizing NN_DESCENT to build knn graph */
+struct graph_build_params : cuvs::neighbors::cagra::graph_build_params {
+  std::optional<cuvs::neighbors::nn_descent::index_params> nn_descent_params = std::nullopt;
```

Comment:
Does this need to be optional? After we move the `construct_index_with_dataset` member out, we can just use the `std::variant` to keep an `nn_descent::index_params` struct optionally.